### PR TITLE
    Path dependent branch folding

### PR DESCRIPTION
**Pull request by meg-gupta**
This optimization redirects branches to their eventual destination, when we can prove the result of all indermediate branches, and is legal to skip all the intermediate instructions.
    Example :

    ```
    function foo(x) {
      var testValue = 100;
      if (x > 100) {
        testValue = 0;  // At this block we can directly redirect the branch to return block, instead of doing following test again
      }
      if (testValue != 0) {
        print(testValue);
      }

    ```
    This helps in efficient merge of valueInfos, so that out global optimizer can aggressively work on blocks, for which it was previously limited due to conservative merge of valueInfos.
    This is the first part of the optimization, a second implementation supporting additional instructions.

### CODE REVIEW

File: lib/Backend/FlowGraph.cpp [hunk 1]
```diff
@@ -4431,6 +4431,311 @@ BasicBlock::CleanUpValueMaps()
...
+            if (!instr->GetDst()->IsRegOpnd()) // complex dstOpnd, stop.
+            {
+                return lastBranchTarget;
+            }
+            IR::RegOpnd *dst = instr->GetDst()->AsRegOpnd();
+            if (currentLabel)
+            {
+                BasicBlock * currentBlock = currentLabel->GetBasicBlock();
+                // If the dstOpnd is defined and used in the same block, it is legal, but if the successor blocks use it, stop.
+                if (currentBlock->succUses && currentBlock->succUses->Test(dst->GetStackSym()->m_id))
```
Comment (meg-gupta) [inline lib/Backend/FlowGraph.cpp]: We need isDead outside the currentBlock, isDead on the Opnd is too strict, so I have added a new bit vector which is upwardExposedUses of the successor blocks

**Comment by meg-gupta** [comment]
TODO : 
- Allow InlineeEnd and LdFld
- Allow optimization in oopjit

File: lib/Backend/GlobOpt.cpp [hunk 2]
```diff
@@ -470,6 +470,7 @@ GlobOpt::OptBlock(BasicBlock *block)
     PrepareLoopArrayCheckHoist();
 
     block->MergePredBlocksValueMaps(this);
+    this->currentBlock->PathDepBranchFolding(this);
```
Comment (sigatrev) [inline lib/Backend/GlobOpt.cpp]: nit: if `block` and `this->currentBlock` are the same here, we should try to be consistent with which one we use.

File: lib/Backend/GlobOpt.cpp [hunk 3]
```diff
@@ -6417,6 +6420,38 @@ GlobOpt::GetConstantVar(IR::Opnd *opnd, Value *val)
     return nullptr;
 }
 
+Js::Var
+GlobOpt::GetConstantVar(IR::Opnd *opnd)
```
Comment (rajatd) [inline lib/Backend/GlobOpt.cpp]: >GetConstantVar [](start = 9, length = 14)

nit: can we refactor the logic of this function into the GetConstantVar above. and pass nullptr from places where you call this function?

File: lib/Backend/GlobOpt.cpp [hunk 4]
```diff
@@ -6417,6 +6420,38 @@ GlobOpt::GetConstantVar(IR::Opnd *opnd, Value *val)
...
+        {
+            return nullptr;
+        }
+        return GetConstantVar(defInstr->GetSrc1());
+    }
+
+    return nullptr;
+}
+
 bool BoolAndIntStaticAndTypeMismatch(Value* src1Val, Value* src2Val, Js::Var src1Var, Js::Var src2Var)
```
Comment (rajatd) [inline lib/Backend/GlobOpt.cpp]: >BoolAndIntStaticAndTypeMismatch [](start = 5, length = 31)

unused?

File: lib/Backend/FlowGraph.cpp [hunk 5]
```diff
@@ -4431,6 +4431,311 @@ BasicBlock::CleanUpValueMaps()
...
+            currentLabel = instr->AsLabelInstr();
+        }
+        if (instr->IsBranchInstr())
+        {
+            IR::BranchInstr* branch = instr->AsBranchInstr();
+            IR::LabelInstr* branchTarget = nullptr;
+            if (branch->IsUnconditional())
+            {
+                branchTarget = branch->GetTarget();
+                if (!branchTarget)
```
Comment (rajatd) [inline lib/Backend/FlowGraph.cpp]: >branchTarget [](start = 21, length = 12)

In what case would an unconditional branch not have a target?

File: lib/Backend/FlowGraph.cpp [hunk 6]
```diff
@@ -4431,6 +4431,311 @@ BasicBlock::CleanUpValueMaps()
...
+                    return lastBranchTarget != nullptr;
+                }
+                if (branchTarget && branchTarget->m_isLoopTop)
+                {
+                    return lastBranchTarget != nullptr;
+                }
+            }
+            else
+            {
+                if (branch->GetTarget()->m_isLoopTop)
```
Comment (rajatd) [inline lib/Backend/FlowGraph.cpp]: >branch->GetTarget()->m_isLoopTop [](start = 20, length = 32)

nit: could move this check before checking the conditionality of the branch

File: lib/Backend/FlowGraph.cpp (hunk 5 above)
Comment (meg-gupta) [reply lib/Backend/FlowGraph.cpp]: For MultiBr

File: lib/Backend/FlowGraph.cpp (hunk 6 above)
Comment (meg-gupta) [reply lib/Backend/FlowGraph.cpp]: GetTarget() can be nullptr

Comment (meg-gupta) [reply lib/Backend/FlowGraph.cpp]: Oh ya, I can move inspite of that.

File: lib/Backend/GlobOpt.cpp (hunk 4 above)
Comment (meg-gupta) [reply lib/Backend/GlobOpt.cpp]: Thanks for catching, it is an early exit check for a few branch types.

**Comment by meg-gupta** [comment]
Fixed most suggestions.

File: lib/Backend/FlowGraph.cpp [hunk 7]
```diff
@@ -4434,6 +4434,312 @@ BasicBlock::CleanUpValueMaps()
...
+            else
+            {
+                src2Val = *src2ValPtr;
+                if (!src2Val)
+                {
+                    return nullptr;
+                }
+            }
+        }
+        src2Var = globOpt->GetConstantVar(branch->GetSrc2());
```
Comment (rajatd) [inline lib/Backend/FlowGraph.cpp]: >GetConstantVar [](start = 27, length = 14)

What's the reason for calling separate overloads of GetConstantVar for src1 and src2? Can't we call the one that takes both Opnd* and Value* for both?

File: lib/Backend/FlowGraph.cpp [hunk 8]
```diff
@@ -4434,6 +4434,312 @@ BasicBlock::CleanUpValueMaps()
...
+                    return nullptr;
+                }
+            }
+            else
+            {
+                src2Val = *src2ValPtr;
+                if (!src2Val)
+                {
+                    return nullptr;
+                }
```
Comment (rajatd) [inline lib/Backend/FlowGraph.cpp]: nit: these checks are identical for src1 and src2. These should ideally be captured in a lambda or a separate method.

File: lib/Backend/FlowGraph.cpp (hunk 7 above)
Comment (meg-gupta) [reply lib/Backend/FlowGraph.cpp]: Fixed, both use the same function now.

**Comment by meg-gupta** [comment]
Thanks @rajatd and @sigatrev for reviewing.

**Pull request merged**